src/checkpoint/dispatch/vrt/virtual_serialize.h

```diff
@@ -132,72 +66,67 @@ void virtualSerialize(T*& base, SerializerT& s) {
+ * and serializing what the pointer points to.
  */
-template <typename T, typename _enabled = void>
-struct StaticAllocateConstruct;
+template <typename T, typename SerializerT, typename _enabled = void>
+struct SerializeAsVirtualIfNeeded;
 
-template <typename T>
-struct StaticAllocateConstruct<
+template <typename T, typename SerializerT>
+struct SerializeAsVirtualIfNeeded<
   T,
-  typename std::enable_if_t<VirtualSerializeTraits<T>::has_not_virtual_serialize>
+  SerializerT,
+  typename std::enable_if_t<
+    dispatch::vrt::VirtualSerializeTraits<T>::has_not_virtual_serialize
+  >
 > {
-  static T* apply() {
+  static void apply(SerializerT& s, T*& target) {
+    // no type idx needed in this case, static construction in default case
     auto t = std::allocator<T>{}.allocate(1);
-    return dispatch::Reconstructor<T>::construct(t);
+    target = dispatch::Reconstructor<T>::construct(t);
   }
 };
 
-template <typename T>
-struct StaticAllocateConstruct<
+template <typename T, typename SerializerT>
+struct SerializeAsVirtualIfNeeded<
   T,
-  typename std::enable_if_t<VirtualSerializeTraits<T>::has_virtual_serialize>
+  SerializerT,
+  typename std::enable_if_t<
+    dispatch::vrt::VirtualSerializeTraits<T>::has_virtual_serialize
+  >
 > {
-  static T* apply() {
-    checkpointAssert(false, "This code should be unreachable");
-    return nullptr;
-  }
-};
+  static void apply(SerializerT& s, T*& target) {
+    using dispatch::vrt::TypeIdx;
 
-}}} /* end namespace checkpoint::dispatch::vrt */
+    TypeIdx entry = dispatch::vrt::no_type_idx;
 
-namespace checkpoint {
-
-template <typename SerializerT, typename T>
-void serializeAllocatePointer(SerializerT& s, T*& target) {
-  using dispatch::vrt::TypeIdx;
-  using dispatch::vrt::OptionalTypeToken;
-  using dispatch::vrt::VirtualSerializeTraits;
-  using dispatch::vrt::RegisterIfNeeded;
-  using dispatch::vrt::StaticAllocateConstruct;
-
-  TypeIdx entry = dispatch::vrt::no_type_idx;
-
-  if (not s.isUnpacking()) {
-    entry = RegisterIfNeeded<T>::apply(target);
-  }
+    if (not s.isUnpacking()) {
+      entry = target->_checkpointDynamicTypeIndex();
+    }
 
-  OptionalTypeToken token(VirtualSerializeTraits<T>::has_virtual_serialize, entry);
-  s | token;
+    s | entry;
 
-  if (s.isUnpacking()) {
-    if (token.isVirtual()) {
+    if (s.isUnpacking()) {
       // use type idx here, registration needed for proper type re-construction
-      auto idx = token.getTypeIdx();
-      auto t = dispatch::vrt::objregistry::getObjAllocate<T>(idx)();
-      target = dispatch::vrt::objregistry::getObjConstruct<T>(idx)(t);
-    } else {
-      // no type idx needed in this case, static construction in default case
-      // we can't directly call because of bad mix between SFINAE and static_assert
-      target = StaticAllocateConstruct<T>::apply();
+      auto t = dispatch::vrt::objregistry::getObjAllocate<T>(entry)();
+      target = dispatch::vrt::objregistry::getObjConstruct<T>(entry)(t);
```

Comment:
Looking at this, might a slight refactoring to just give the registry functions that call the function objects and return their results be cleaner?

I.e. 
```
auto t = dispatch::vrt::objregistry::allocate<T>(entry);
target = dispatch::vrt::objregistry::construct<T>(entry, t);
```

Reply:
Sure, why not!

Reply:
Done